graph.cpp

```diff
@@ -0,0 +1,194 @@
+  incoming->insert(from);
+}
+
+void Graph::Node::insertOutgoing(int64 to) {
+  outgoing->insert(to);
+}
+
+void Graph::Node::eraseIncoming(int64 from) {
+  incoming->erase(from);
+}
+
+void Graph::Node::eraseOutgoing(int64 to) {
+  outgoing->erase(to);
+}
+
+bool Graph::Node::containsEdgeTo(int64 to) {
+  // apparently the standard c++ way to check if something is in range
+  // call this too many times to not make this a method...
+  return outgoing->find(to) != outgoing->end();
+}
+
+bool Graph::Node::containsEdgeFrom(int64 from) {
+  return incoming->find(from) != incoming->end();
+}
+
+bool Graph::Contains(int64 nodeid) {
+  return nodemap.find(nodeid) != nodemap.end();
+}
+
+Graph::Node Graph::Node::DeepCopy() {
+  // create a new node
+  auto result = Node();
+  // go through all the incoming and outgoing edges and insert them in the
+  // result. That's all there is to the state of the node, since we are not
+  // storing the id...
+  for(auto v : *outgoing) {
+    result.outgoing->insert(v);
+  }
+  for(auto v : *incoming) {
+    result.incoming->insert(v);
+  }
+  return result;
+}
+
+int64 Graph::AddNode() {
+  // The counter is keeping track of the ids that have been assigned so far.
+  static int64 id_counter = 0;
+  static int64 max_int64 = std::numeric_limits<int64>::max();
+  
+  // The only reason we are doing mod maxint64 and checking that
+  // the id already exists is on the off chance that about
+  // maxint64 addNodes might have already happened on this graph.
+  // Probably not going to happen, but good karma, right?
+  
+  while(Contains(id_counter)) id_counter = (id_counter % max_int64) + 1;
+  nodemap.insert({id_counter, Node()});
+  return id_counter;
+}
+
+int64 Graph::AddNode(int64 id) {
+  if(!Contains(id)) {
+    nodemap.insert({id, Node()});
+  }
+  return id;
+}
+
+int64 Graph::Count() {
+  // an invariant we mantain is that number of actual nodes == number of nodes in the map
+  return nodemap.size();
+}
+
+void Graph::Connect(int64 from, int64 to) {
+  // check if both nodes are in the graph
+  if(Contains(from) && Contains(to)) {
+    // insert the edge
+    nodemap.at(from).insertOutgoing(to);
+    nodemap.at(to).insertIncoming(from);
+  }
+}
+
+void Graph::Disconnect(int64 from, int64 to) {
+  // check if both nodes are in the graph
+  if(Contains(from) && Contains(to)) {
+    // delete the edge
+    nodemap.at(from).eraseOutgoing(to);
+    nodemap.at(to).eraseIncoming(from);
+  }
+}
+
+bool Graph::IsConnected(int64 from, int64 to) {
+  bool result = false;
+  if(Contains(from)) {
+    result = nodemap.at(from).containsEdgeTo(to);
+  }
+  return result;
+}
+
+Graph Graph::DeepCopy() {
+  // make a new graph and copy over all nodes. The nodes have their own deep
+  // copy method
+  auto result = Graph();
+  for(auto& kv : nodemap) {
+    // k -> id, v -> node
+    result.nodemap.insert({kv.first, kv.second.DeepCopy()});
+  }
+  return result;
+}
+
+void Graph::Reverse(Graph* graph_to_reverse) {
```

Comment:
Beautiful, dude. I see why you preferred using a pointer for these lists now. I'm going to revise my comment in the header file now.